cheroot/connections.py

```diff
@@ -99,76 +99,87 @@ def expire(self):
         # find any connections still registered with the selector
         # that have not been active recently enough.
         threshold = time.time() - self.server.timeout
-        timed_out_connections = (
+        timed_out_connections = [
             (sock_fd, conn)
             for _, (_, sock_fd, _, conn) in self._selector.get_map().items()
             if conn != self.server and conn.last_used < threshold
-        )
+        ]
         for sock_fd, conn in timed_out_connections:
             self._selector.unregister(sock_fd)
             conn.close()
 
-    def get_conn(self):  # noqa: C901  # FIXME
-        """Return a HTTPConnection object which is ready to be handled.
+    def stop(self):
+        """Stop the selector loop in run() synchronously.
 
-        A connection returned by this method should be ready for a worker
-        to handle it. If there are no connections ready, None will be
-        returned.
+        May take up to half a second.
+        """
+        self._stop_requested = True
+        while self._serving:
+            time.sleep(0.01)
 
-        Any connection returned by this method will need to be `put`
-        back if it should be examined again for another request.
+    def run(self):
+        """Run the connections selector indefinitely.
 
-        Returns:
-            cheroot.server.HTTPConnection instance, or None.
+        Connections that are ready to process are submitted via
+        self.server.process_conn()
 
-        """
-        # return a readable connection if any exist
-        with suppress(IndexError):
-            return self._readable_conns.popleft()
-
-        # Will require a select call.
-        try:
-            # The timeout value impacts performance and should be carefully
-            # chosen. Ref:
-            # github.com/cherrypy/cheroot/issues/305#issuecomment-663985165
-            rlist = [
-                key for key, _
-                in self._selector.select(timeout=0.01)
-            ]
-        except OSError:
-            # Mark any connection which no longer appears valid
-            for _, key in self._selector.get_map().items():
-                # If the server socket is invalid, we'll just ignore it and
-                # wait to be shutdown.
-                if key.data == self.server:
-                    continue
+        Connections submitted for processing must be `put()`
+        back if it should be examined again for another request.
 
-                try:
-                    os.fstat(key.fd)
-                except OSError:
-                    # Socket is invalid, close the connection
-                    self._selector.unregister(key.fd)
+        Can be shut down by calling `stop()`.
+        """
+        self._serving = True
+        expiration_interval = 0.5
```

Comment:
this value is a bit arbitrary, but seemed reasonable enough. let me know if you'd like to see this configurable in some other way.